File: Sources/HaulCore/Download/Aria2c.swift
import Foundation

public enum Aria2c {
    public static func download(_ url: String, to path: String, config: DownloadConfig) async throws {
        let args = arguments(url, to: path, config: config)
        Log.debug("aria2c \(args.joined(separator: " "))")
        try await Shell.run(config.aria2cPath, args, echo: false)
        if FileManager.default.fileExists(atPath: path + ".aria2") || !FileManager.default.fileExists(atPath: path) {
            throw HaulError("aria2c failed")
        }
        Log.plain()
    }

    /// Same headers as the built-in downloader: the site's own when given, else the bilibili defaults.
    static func arguments(_ url: String, to path: String, config: DownloadConfig) -> [String] {
        var args = ["--auto-file-renaming=false", "--download-result=hide", "--allow-overwrite=true", "--console-log-level=warn",
                    "-x16", "-s16", "-j16", "-k5M"]
        if config.headers.isEmpty {
            if !url.contains("platform=android_tv_yst") && !url.contains("platform=android") {
                args.append("--header=Referer: https://www.bilibili.com")
            }
            args.append("--header=User-Agent: Mozilla/5.0")
        }
        for (k, v) in config.headers.sorted(by: { $0.key < $1.key }) { args.append("--header=\(k): \(v)") }
        if !config.cookie.isEmpty { args.append("--header=Cookie: \(config.cookie)") }
        args += splitArguments(config.aria2cArgs)
        let dir = (path as NSString).deletingLastPathComponent
        args += [url, "-d", dir.isEmpty ? "." : dir, "-o", (path as NSString).lastPathComponent]
        return args
    }

    /// Shell-style word splitting with single/double quotes, for `--aria2c-args`.
    public static func splitArguments(_ text: String) -> [String] {
        var out: [String] = []
        var current = ""
        var quote: Character?
        var hasToken = false
        var escaped = false
        for ch in text {
            if escaped { current.append(ch); escaped = false; continue }
            if ch == "\\" && quote != "'" { escaped = true; continue }
            if let q = quote {
                if ch == q { quote = nil } else { current.append(ch) }
                continue
            }
            if ch == "\"" || ch == "'" { quote = ch; hasToken = true; continue }
            if ch.isWhitespace {
                if hasToken { out.append(current); current = ""; hasToken = false }
                continue
            }
            current.append(ch)
            hasToken = true
        }
        if hasToken { out.append(current) }
        return out
    }
}

File: internal/fetch/aria2c.go
package fetch

import (
	"context"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"unicode"

	"github.com/ac1982/haul/internal/errs"
	"github.com/ac1982/haul/internal/media"
	"github.com/ac1982/haul/internal/shell"
)

// Aria2c downloads through aria2c. Resources that allow only closed ranges one at a time go to Fallback instead:
// aria2c cannot be held to that.
type Aria2c struct {
	Path     string
	Args     []string
	Fallback Fetcher
}

// Fetch runs aria2c for res.
func (a *Aria2c) Fetch(ctx context.Context, res media.Resource, dst string, p Progress) error {
	if res.Policy == media.Sequential && a.Fallback != nil {
		return a.Fallback.Fetch(ctx, res, dst, p)
	}
	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}
	result, err := shell.Run(ctx, a.Path, a.Arguments(res, dst), shell.Options{Echo: true})
	if err != nil {
		return err
	}
	if _, err := os.Stat(dst + ".aria2"); result.Status != 0 || err == nil {
		return errs.New("aria2c failed (exit code %d)", result.Status)
	}
	if _, err := os.Stat(dst); err != nil {
		return errs.New("aria2c wrote nothing")
	}
	return nil
}

// Arguments is the aria2c command line: the resource's headers, sorted, then the extra arguments.
func (a *Aria2c) Arguments(res media.Resource, dst string) []string {
	args := []string{"--auto-file-renaming=false", "--download-result=hide", "--allow-overwrite=true",
		"--console-log-level=warn", "-x16", "-s16", "-j16", "-k5M"}
	names := make([]string, 0, len(res.Header))
	for k := range res.Header {
		names = append(names, k)
	}
	sort.Strings(names)
	for _, k := range names {
		for _, v := range res.Header[k] {
			args = append(args, "--header="+k+": "+v)
		}
	}
	args = append(args, a.Args...)
	return append(args, res.URL, "-d", filepath.Dir(dst), "-o", filepath.Base(dst))
}

// SplitArgs is shell-style word splitting with quotes and backslashes, for --aria2c-args.
func SplitArgs(text string) []string {
	var out []string
	var cur strings.Builder
	var quote rune
	hasToken, escaped := false, false
	for _, ch := range text {
		switch {
		case escaped:
			cur.WriteRune(ch)
			escaped = false
		case ch == '\\' && quote != '\'':
			escaped = true
		case quote != 0:
			if ch == quote {
				quote = 0
			} else {
				cur.WriteRune(ch)
			}
		case ch == '"' || ch == '\'':
			quote = ch
			hasToken = true
		case unicode.IsSpace(ch):
			if hasToken {
				out = append(out, cur.String())
				cur.Reset()
				hasToken = false
			}
		default:
			cur.WriteRune(ch)
			hasToken = true
		}
	}
	if hasToken {
		out = append(out, cur.String())
	}
	return out
}
